# Name: Jake Mavrides

### (Optional) Feedback/where I ran into trouble/etc. - 

In this assignment, you will be taking a closer look at the Spam Classifier we discussed in lecture.

In [ ]:
# plotting
import matplotlib.pyplot as plt
# pandas
import pandas as pd
# sklearn for accessing ML operations and SVM algorithm
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn import svm

Spam dataset courtesy of UCI :D https://www.kaggle.com/uciml/sms-spam-collection-dataset/

Additional Notes and References:
1. https://blog.logrocket.com/email-spam-detector-python-machine-learning/
2. https://github.com/tejank10/Spam-or-Ham/blob/master/spam_ham.ipynb

In [ ]:
# load data into a pandas dataframe
spam_df = pd.read_csv('spam.csv')
# let's make it pretty...
spam_df.drop(['Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], axis = 1, inplace = True)
spam_df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [ ]:
# CODE SECTION 1
# rename the v1 and v2 columns to 'label' and 'contents' respectively
spam_df.rename(columns={'v1':'label', 'v2':'contents'}, inplace=True)

In [ ]:
# CODE SECTION 2
# extract contents and ensure they are the same size
contents = spam_df['contents']
label = spam_df['label']
assert(contents.shape == label.shape)
label # should only show ham and spam in column

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: label, Length: 5572, dtype: object

In [ ]:
# CODE SECTION 3
# train multiple SVM ML models using increments of 10% of test data from 9% to 99% (9%, 19%, 29%, ..., 99%)
# after each iteration, print out the resulting statistics for test size used and accuracy of the model
# e.g. Test size percentage: ??? and Accuracy: ???
# hint: see the lecture notes from this week underneath the diagram of the SVM model
for i in range(1,11): #Loop 10 times
  testsize = -.01 + (.10 * i)
  contents_train,contents_test,label_train,label_test = train_test_split(contents,label,test_size=testsize)
  cv = CountVectorizer()
  features=cv.fit_transform(contents_train)
  model = svm.SVC()
  model.fit(features,label_train)
  features_test = cv.transform(contents_test)
  print("Test size percentage: " + str(round(testsize, 2)))
  print("Accuracy: {}".format(model.score(features_test,label_test)))

Test size percentage: 0.09
Accuracy: 0.9701195219123506
Test size percentage: 0.19
Accuracy: 0.9754485363550519
Test size percentage: 0.29
Accuracy: 0.9801980198019802
Test size percentage: 0.39
Accuracy: 0.9825206991720331
Test size percentage: 0.49
Accuracy: 0.9787623581105822
Test size percentage: 0.59
Accuracy: 0.9717153284671532
Test size percentage: 0.69
Accuracy: 0.9677503250975292
Test size percentage: 0.79
Accuracy: 0.9631985461154021
Test size percentage: 0.89
Accuracy: 0.9245967741935484
Test size percentage: 0.99
Accuracy: 0.870763095885445


### Q1: Based on the previous section, do you broadly see a relationship between test size and accuracy?
I definitely see a relationship, but maybe not a definitive/very broad one.

### Q2: Based on your answer in Q1, why do you think this is?
I would've expected the larger and larger sample training size to yield higher and higher accuracy, but after seeing the results this is not the case. I suspect that this is because when we increase the sample training size, we're decreasing the testing size afterwards. In our example the balance appears to be closest to be around size=.39

### Q3: The process of breaking a stream of text up into words, phrases, symbols, or other meaningful elements called tokens is usually referred to as ???
Tokenization

### Q4: The ??? function from the sklearn package was used to train the model with a percentage of the dataset.
train_test_split

### Q5: The SVM model is commonly used as a (1) supervised/unsupervised machine learning algorithm. For instance, in the Spam Classification problem, the data is (2) labeled/unlabeled.
(1): Supervised
(2): Labeled